File: appcore/src/main/java/mobi/cangol/mobile/soap/SoapClient.java
/**
 * Copyright (c) 2013 Cangol.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package mobi.cangol.mobile.soap;

import android.content.Context;
import android.util.Log;

import org.ksoap2.SoapEnvelope;
import org.ksoap2.serialization.SoapObject;
import org.ksoap2.serialization.SoapSerializationEnvelope;
import org.ksoap2.transport.HttpTransportSE;
import org.kxml2.kdom.Element;
import org.kxml2.kdom.Node;

import java.lang.ref.WeakReference;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.WeakHashMap;
import java.util.concurrent.Future;

import mobi.cangol.mobile.service.PoolManager;

/**
 *  SoapClient.java 使用此类需要ksoap2-android-assembly-3.0.0-jar-with-dependencies.jar
 * @author Cangol
 */
public class SoapClient {
    private static final String TAG = "SoapRequest";
    private final static int TIMEOUT = 20 * 1000;
    private final Map<Context, List<WeakReference<Future<?>>>> requestMap;
    private SoapSerializationEnvelope envelope;
    private HttpTransportSE ht;
    private PoolManager.Pool threadPool;

    public SoapClient() {
        threadPool = PoolManager.buildPool(TAG, 3);
        requestMap = new WeakHashMap<Context, List<WeakReference<Future<?>>>>();
        envelope = new SoapSerializationEnvelope(SoapEnvelope.VER12);
        envelope.dotNet = true;
    }

    /**
     * 添加header
     * @param namespace
     * @param authheader
     * @param headers
     */
    public void addHeader(String namespace, String authheader, HashMap<String, String> headers) {
        if (authheader != null && headers != null) {
            envelope.headerOut = new Element[1];
            envelope.headerOut[0] = buildAuthHeader(namespace, authheader, headers);
        }
    }

    /**
     * 执行请求
     * @param context
     * @param url
     * @param namespace
     * @param action
     * @param params
     * @param responseHandler
     */
    public void send(Context context, String url, String namespace, String action, HashMap<String, String> params, SoapResponseHandler responseHandler) {

        if (params != null) {
            StringBuilder paramsStr = new StringBuilder(url);
            paramsStr.append('/')
                    .append(action);
            if (params.size() > 0) {
                paramsStr.append('?');
            }
            SoapObject rpc = new SoapObject(namespace, action);
            for (Map.Entry<String, String> entry : params.entrySet()) {
                rpc.addProperty(entry.getKey(), entry.getValue());
                paramsStr.append(entry.getKey())
                        .append('=')
                        .append(entry.getValue())
                        .append('&');
            }
            Log.d(TAG, "sendRequest " + paramsStr.toString());
            envelope.bodyOut = rpc;
        }
        ht = new HttpTransportSE(url, TIMEOUT);
        sendRequest(ht, envelope, namespace, responseHandler, context);
    }

    /**
     * 构建auth header
     * @param namespace
     * @param authheader
     * @param params
     * @return
     */
    private Element buildAuthHeader(String namespace, String authheader, HashMap<String, String> params) {
        Element header = new Element().createElement(namespace, authheader);
        for (Map.Entry<String, String> entry : params.entrySet()) {
            Element element = new Element().createElement(namespace, entry.getKey());
            element.addChild(Node.TEXT, entry.getValue());
            header.addChild(Node.ELEMENT, element);
        }
        return header;
    }

    /**
     * 取消请求
     * @param context
     * @param mayInterruptIfRunning
     */
    public void cancelRequests(Context context, boolean mayInterruptIfRunning) {
        List<WeakReference<Future<?>>> requestList = requestMap.get(context);
        if (requestList != null) {
            for (WeakReference<Future<?>> requestRef : requestList) {
                Future<?> request = requestRef.get();
                if (request != null) {
                    request.cancel(mayInterruptIfRunning);
                }
            }
        }
        requestMap.remove(context);
    }

    /**
     * 发生请求
     * @param ht
     * @param envelope
     * @param namespace
     * @param responseHandler
     * @param context
     */
    protected void sendRequest(HttpTransportSE ht, SoapSerializationEnvelope envelope,
                               String namespace, SoapResponseHandler responseHandler, Context context) {

        Future<?> request = threadPool.submit(new SoapRequest(ht, envelope, namespace, responseHandler));

        if (context != null) {
            // Add request to request map
            List<WeakReference<Future<?>>> requestList = requestMap.get(context);
            if (requestList == null) {
                requestList = new LinkedList<WeakReference<Future<?>>>();
                requestMap.put(context, requestList);
            }
            requestList.add(new WeakReference<Future<?>>(request));
        }
    }
}

File: appcore/src/main/java/mobi/cangol/mobile/soap/SoapClient.kt
/**
 * Copyright (c) 2013 Cangol.
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package mobi.cangol.mobile.soap

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.Message
import mobi.cangol.mobile.Task
import mobi.cangol.mobile.logging.Log
import mobi.cangol.mobile.service.PoolManager
import mobi.cangol.mobile.socket.SocketClient
import org.ksoap2.SoapEnvelope
import org.ksoap2.serialization.SoapObject
import org.ksoap2.serialization.SoapSerializationEnvelope
import org.ksoap2.transport.HttpTransportSE
import org.kxml2.kdom.Element
import org.kxml2.kdom.Node
import java.lang.ref.SoftReference
import java.lang.ref.WeakReference
import java.util.*
import java.util.concurrent.Future

/**
 * SoapClient.java 使用此类需要ksoap2-android-assembly-3.0.0-jar-with-dependencies.jar
 *
 * @author Cangol
 */
class SoapClient {
    private val requestMap= mutableMapOf<Context, MutableList<WeakReference<Future<*>>>>()
    private val envelope=SoapSerializationEnvelope(SoapEnvelope.VER12)
    private val threadPool: PoolManager.Pool = PoolManager.buildPool("SoapRequest", 3)

    init {
        envelope.dotNet = true
    }

    /**
     * 添加header
     *
     * @param namespace
     * @param authheader
     * @param headers
     */
    fun addHeader(namespace: String, authHeader: String?, headers: Map<String, String>?) {
        if (authHeader != null && headers != null) {
            envelope.headerOut = arrayOfNulls(1)
            envelope.headerOut[0] = buildAuthHeader(namespace, authHeader, headers)
        }
    }

    /**
     * 执行请求
     *
     * @param context
     * @param url
     * @param namespace
     * @param action
     * @param params
     * @param responseHandler
     */
    fun send(context: Context, url: String, namespace: String, action: String, params: Map<String, String>?, responseHandler: SoapResponseHandler) {

        if (params != null) {
            val paramsStr = StringBuilder(url)
            paramsStr.append('/')
                    .append(action)
            if (params.isNotEmpty()) {
                paramsStr.append('?')
            }
            val rpc = SoapObject(namespace, action)
            for ((key, value) in params) {
                rpc.addProperty(key, value)
                paramsStr.append(key)
                        .append('=')
                        .append(value)
                        .append('&')
            }
            Log.d("sendRequest $paramsStr")
            envelope.bodyOut = rpc
        }
        sendRequest(HttpTransportSE(url, 20 * 1000), envelope, namespace, responseHandler, context)
    }

    /**
     * 构建auth header
     *
     * @param namespace
     * @param authHeader
     * @param params
     * @return
     */
    private fun buildAuthHeader(namespace: String, authHeader: String, params: Map<String, String>): Element {
        val header = Element().createElement(namespace, authHeader)
        for ((key, value) in params) {
            val element = Element().createElement(namespace, key)
            element.addChild(Node.TEXT, value)
            header.addChild(Node.ELEMENT, element)
        }
        return header
    }

    /**
     * 取消请求
     *
     * @param context
     * @param mayInterruptIfRunning
     */
    fun cancelRequests(context: Context, mayInterruptIfRunning: Boolean) {
        val requestList = requestMap[context]
        if (requestList != null) {
            for (requestRef in requestList) {
                val request = requestRef.get()
                request?.cancel(mayInterruptIfRunning)
            }
        }
        requestMap.remove(context)
    }

    /**
     * 发生请求
     *
     * @param ht
     * @param envelope
     * @param namespace
     * @param responseHandler
     * @param context
     */
    private fun sendRequest(ht: HttpTransportSE, envelope: SoapSerializationEnvelope,
                              namespace: String, responseHandler: SoapResponseHandler, context: Context?) {

        val request = threadPool.submit(SoapRequest(ht, envelope, namespace, responseHandler))

        if (context != null) {
            // Add request to request map
            var requestList: MutableList<WeakReference<Future<*>>>? = requestMap[context]
            if (requestList == null) {
                requestList = LinkedList()
                requestMap[context] = requestList
            }
            requestList.add(WeakReference(request))
        }
    }
}
